### DISCUSSION TITLE
Processing values from signals (match_signal closure)

### DISCUSSION THREAD

**Discussion by debug-richard** [diwic/dbus-rs#393] [Q&A] [answered]
I am currently working with the org.freedesktop.PackageKit dbus api.
The API is asynchronous and returns a signal for each package to install.
I have used the https://github.com/diwic/dbus-rs/blob/master/dbus/examples/match_signal.rs example to catch the signals and it worked.
However, I am now wondering how to get the "h: ComExampleDbustestHelloHappened" object out of the closure/function and store it in a Vec for further processing.
I tried a few things, but it didn't work. 
Somehow it feels that dbus-rs is missing this use case.

**Comment by debug-richard** [comment, marked as answer]
After all I found a solution to get the data out of the closure using Arc and Mutex:

```rust
// Note: Untested modification of the example
let result: Arc<Mutex<Vec<ComExampleDbustestHelloHappened>>> = Arc::new(Mutex::new(Vec::new()));
let result_2 = result.clone();
let _id = proxy.match_signal(|h: ComExampleDbustestHelloHappened, _: &Connection, _: &Message| {
    println!("Hello happened from sender: {}", h.sender);
    result_2.lock().unwrap().push(h);
    true
});
println!("{:?}", result);

```

**Reply by diwic** [reply]
This is more of a question about boxed closures than about dbus, but this is one way of doing it. You can also use [channels](https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html).
